### PR TITLE
win32ui: Fix sys.argv Py3.8+ (#1620). More startup options.

### PR DESCRIPTION
**Pull request by kxrob**
commit 2 + 3 : more startup options. 3 is necessary so that with start_pythonwin.py several dialogs don't go mad with automatic button pressing. The "fixup!" commit is just waiting for next or final rebase & force push.

(1. commit moved to extra PR)
(Rest of commits already split off and done)

### CODE REVIEW

**Comment by kxrob** [comment]
( The commit `postinstall: Fix registry path for PyWin32.chm when 32bit (#1621)` loosly fits here - and CI doesn't fail completely here :)  )

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -36,7 +36,8 @@ jobs:
       run: pip install --upgrade setuptools wheel
 
     - name: Set Python user directory
-      run: echo "::set-env name=USER_DIR::$(python -c 'import os,site;print(os.path.dirname(site.USER_SITE))', end='')"
+      run: python -c "import os,site;open(os.environ['GITHUB_ENV'], 'a').write('USER_DIR=%s\n' % os.path.dirname(site.USER_SITE))"
+      ## run: echo "USER_DIR=$(python -c 'import sys,os,site;sys.stdout.write(os.path.dirname(site.USER_SITE))')" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
```
Comment (mhammond) [inline .github/workflows/main.yml:40]: Can you please remove this commented line? (It would actually be great if this was its own PR, so it could land immediately)

File: Pythonwin/pywin/framework/startup.py [hunk 2]
```diff
@@ -20,8 +20,8 @@
 	# sys.argv and it will be done here. ( This was a workaround for a bug in
 	# win32ui but is retained for other situations. )
 	argv = win32api.CommandLineToArgv(win32api.GetCommandLine())
-	sys.argv = argv[1:]
-	if os.getcwd() not in sys.path and '.' not in sys.path:
+	sys.argv = argv[1:] or ['']
```
Comment (mhammond) [inline Pythonwin/pywin/framework/startup.py:23]: why this change?

File: Pythonwin/win32uimodule.cpp [hunk 3]
```diff
@@ -2614,7 +2614,16 @@ extern "C" PYW_EXPORT BOOL Win32uiApplicationInit(Win32uiHostGlue *pGlue, TCHAR
         PySys_SetArgv(__argc - 1, __targv + 1);
 #else
     PyInit_win32ui();
-    if (argv == NULL || !PyList_Check(argv) || !PyList_Size(argv)) {
+    // Decide if we render sys.argv from command line.
+    // PY3.6- Py_Initialize sets sys.argv=NULL .
+    // PY3.7 Py_Initialize or intentional script triggers set sys.argv=[] .
+    // PY3.8+ Py_Initialize sets sys.argv=[''] - cannot be distinguished
```
Comment (mhammond) [inline Pythonwin/win32uimodule.cpp:2620]: sys.argv being exactly `['']` seems unlikely to ever be a "real" argv?

File: pywin32_postinstall.py [hunk 4]
```diff
@@ -215,11 +214,17 @@ def RegisterPythonwin(register=True):
...
     pythonwin_exe = os.path.join(lib_dir, "Pythonwin", "Pythonwin.exe")
     pythonwin_edit_command=pythonwin_exe + ' -edit "%1"'
+    pythonwin_new_command=pythonwin_exe + ' -new "%1"'
+    pythonwin_run_command=pythonwin_exe + ' -run "%1"'
 
     keys_vals = [
         ('Software\\Microsoft\\Windows\\CurrentVersion\\App Paths\\Pythonwin.exe', '', pythonwin_exe),
         ('Software\\Classes\\Python.File\\shell\\Edit with Pythonwin', 'command', pythonwin_edit_command),
         ('Software\\Classes\\Python.NoConFile\\shell\\Edit with Pythonwin', 'command', pythonwin_edit_command),
+        ('Software\\Classes\\Python.File\\shell\\Edit in new Pythonwin', 'command', pythonwin_new_command),
```
Comment (mhammond) [inline pywin32_postinstall.py:224]: I'm not sure these are worthwhile - the more options, the harder it is to find the ones you actually want. And once you have multiple pythonwin's, you've kinda broken the "Edit with Pythonwin" (as you can't control which one). IOW, I'm not sure we should *encourage* multiple pythonwins?

File: Pythonwin/win32uimodule.cpp [hunk 5]
```diff
@@ -2614,7 +2614,16 @@ extern "C" PYW_EXPORT BOOL Win32uiApplicationInit(Win32uiHostGlue *pGlue, TCHAR
...
+    //   from a pre-existing command line setup anymore. So we need to check
+    //   another flag regarding the intended type of invokation, e.g. `cmd`
+    //   (or untangle all that crossover startup + module + app init here)
+    // `cmd` is non-NULL upon Pythonwin.exe / C++ embedded glue startup and
+    //   always NULL during "import win32ui" from normal Python.
+    if (argv == NULL || !PyList_Check(argv) || !PyList_Size(argv) || cmd) {
```
Comment (mhammond) [inline Pythonwin/win32uimodule.cpp:2626]: Is this the only change actually necessary for the PR description?

**Review by mhammond — CHANGES_REQUESTED** [review]
Thanks, and sorry for the delay. There seem to be a whole lot of changes here and it's not clear which are actually fixing the description in the PR - and with the fixup commit I don't really want to land them individually. Could you please split this into 3 PRs - one for the CI fix, one for the argv fix, and one for the enhancements (the latter might need some discussion)

File: .github/workflows/main.yml (hunk 1 above)
Comment (kxrob) [reply .github/workflows/main.yml:40]: Done as new PR #1625 .

File: Pythonwin/pywin/framework/startup.py (hunk 2 above)
Comment (kxrob) [reply Pythonwin/pywin/framework/startup.py:23]: * according Python standard the init'ed sys.argv shall not be a empty list but [''].
* CWD can come in as '' and should be 1st in sys.path (unless a certain script is run in normal python)

File: Pythonwin/win32uimodule.cpp (hunk 3 above)
Comment (kxrob) [reply Pythonwin/win32uimodule.cpp:2620]: Its the normal case when plain python is started without arguments. In 3.8+ now they set that already in Py_Initialize for considering commandline (or alternatives). [] instead of NULL was in 3.7 - one could think this meant "not fully initialized (as it is indeed without commandline inspection or alternative decision) but at least: existing and a list" (that would be reasonable). But they somehow went futher perhaps thinking it was less than perfect ... 
```
C:\...>python3
Python 3.7.8 (tags/v3.7.8:4b47a5b6ba, Jun 28 2020, 08:53:46) [MSC v.1916 64 bit (AMD64)] on win32

Type "help", "copyright", "credits" or "license" for more information.
>>> sys.argv
['']
```

File: pywin32_postinstall.py (hunk 4 above)
Comment (kxrob) [reply pywin32_postinstall.py:224]: I can remove the 2 context commands again from 2nd commit. My motivation was:
* The -run thing lets you comfortably run scripts and inspect / debug the aftermath in PythonWin  (with the extended Debugger etc. to come)
* the -new : I indeed use a 2nd or even 3rd PythonWin quite often as its fast and light-weigth (Actually I use even more context commands for different PY versions).  The -new option is also rather unkown otherwise and one is forced into one instance. Even fatter editors like VSCode etc. can easily have more instances. The distinction is still there: The "Edit with Python" always goes into the one "main" instance. While the secondary -new ones don't listen on DDE at all.
* It may be more easy to remove the commands for a programmer than to know about the alternative startup.

File: Pythonwin/win32uimodule.cpp (hunk 5 above)
Comment (kxrob) [reply Pythonwin/win32uimodule.cpp:2626]: This first one fixes #1620 the Py3.8+ issue. 

The first 3 commits are about startup - 3  is necessary  so that with 2's startup several dialog don't go mad with automatic button pressing. The "fixup!" commit of course is just waiting for next or final rebase & force push.

The 32bit help fix for #1621 commit was just here because of the CI issue -> extra PR?

**Comment by kxrob** [comment]
Part of these patches are already merged via extra PRs. Better close this one - remainders to reappear separately.
